codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -496,8 +518,20 @@ private void incrementPendingBytes(int numBytes) {
              */
             int write(int allowedBytes) {
                 int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
+                int writtenBytes = 0;
+                boolean success = false;
+                try {
+                    needFlush |= payload.write(Math.max(0, allowedBytes));
+                    success = true;
+                } catch (Throwable t) {
+                    payload.error(t);
+                } finally {
+                    if (success && payload.size() == 0) {
+                        pendingWriteQueue.remove();
+                        payload.writeComplete();
+                    }
+                }
+                writtenBytes = before - payload.size();
```

Comment:
@buchgr I think we want to ensure that the flow control windows are updated even if an exception occurs.  Maybe change it around to this:

```java
            int write(int allowedBytes) {
                int before = payload.size();
                int writtenBytes = 0;
                try {
                    try {
                        needFlush |= payload.write(Math.max(0, allowedBytes));
                        if(payload.size() == 0) {
                            pendingWriteQueue.remove();
                            payload.writeComplete();
                        }
                    } catch (Throwable t) {
                        onError(error);
                    }
                } finally {
                    // Make sure we always decrement the flow control windows by the bytes written.
                    writtenBytes = before - payload.size();

                    // TODO: maybe make this try/catch a "decrement" method to help make this more readable?
                    try {
                        connectionState().incrementStreamWindow(-writtenBytes);
                        incrementStreamWindow(-writtenBytes);
                    } catch (Http2Exception e) { // Should never get here since we're decrementing.
                        throw new RuntimeException("Invalid window state when writing frame: " + e.getMessage(), e);
                    }
                    decrementPendingBytes(writtenBytes);
                }
                return writtenBytes;
            }
```